### DISCUSSION TITLE
Using the PublishMassTransitMessage activity from the designer

### DISCUSSION THREAD

**Discussion by webley** [elsa-workflows/elsa-core#2756] [Q&A] [answered]
Are there any samples of how to configure the PublishMassTransitMessage activity from the workflow designer? The editor shows a single text box that can accept literal, JavaScript or Liquid.

![image](https://user-images.githubusercontent.com/9415127/154053786-98ff2adc-f403-482a-a113-4b2a258df6c1.png)

Trying to use JavaScript to return an expression like this:

```
return {
   $type: "MyTest.Application.Events.TestSuccessfulEvent, MyTest.Application",
   Id: 42,
   Name: "Test name"
}
```

results in the error 
```
"exception": {
	"Type": "System.ArgumentException",
	"Message": "Messages types must not be in the System namespace: System.Dynamic.ExpandoObject (Parameter 'T')",
}
```

Defining a JavaScript TypeDefinitionProvider for the specific type allows it to be recognised by Intellisense in the editor. However, trying something like this:

```
var evt = new TestSuccessfulEvent();
evt.Id = 42;
evt.Name = "Test name";
return evt;
```

results in the error
```
"exception": {
	"Type": "Elsa.Exceptions.ExpressionEvaluationException",
	"Message": "Failed to evaluate expression",	
	"InnerException": {
		"Type": "Jint.Runtime.JavaScriptException",
		"Message": "TestSuccessfulEvent is not defined"
	}
}
```

Is it possible to write an expression that returns an instance of a .NET type so MassTransit can publish the message?

**Comment by sfmskywalker** [comment, marked as answer]
Hi @webley 

You certainly can! And in fact, you are very close to the solution with your JS expression and custom `TypeDefinitionProvider`. The only thing missing is a custom handler that configures the Jint engine with access to the type you're trying to instantiate (which will prevent the "TestSuccessfulEvent is not defined" error you received.

Here's an example of such a handler:

```csharp
/// <summary>
/// Configure the JS engine to allow instantiation of message classes.
/// </summary>
public class ConfigureJavaScriptEngine : INotificationHandler<EvaluatingJavaScriptExpression>
{
    public Task Handle(EvaluatingJavaScriptExpression notification, CancellationToken cancellationToken)
    {
        var engine = notification.Engine;
        
        engine.RegisterType<TestSuccessfulEvent>();
        return Task.CompletedTask;
    }
}
```

To register it, make sure you do this from `Startup`:

```csharp
// Register notification handlers from this project. Will register ConfigureJavaScriptEngine.
services.AddNotificationHandlersFrom<Startup>();
```

A working setup can be found in [this updated sample project](https://github.com/elsa-workflows/elsa-core/tree/master/src/samples/aspnet/Elsa.Samples.MassTransitRabbitMq).

With that in place, you can do exactly what you tried to do already:

```javascript
var evt = new TestSuccessfulEvent();
evt.Id = 42;
evt.Name = "Test name";
return evt;
```

**Reply by sfmskywalker** [reply]
UPDATE: You can now also directly configure the engine directly from Program.cs, e.g.:

```csharp
elsa.UseJavaScript(options =>
{
    options.ConfigureEngine(engine =>
    {
        engine.RegisterType<TestSuccessfulEvent>();
    });
})
```

This way, there's no need for implementing a notification handler.

**Comment by pvanroos** [comment]
Are there any examples of publishing a MassTransit message directly from the workflow builder (e.g. builder.PublishMassTransitMessage(...).  The examples only have one where a message is being received.  The publish is happening in the controller.

**Reply by sfmskywalker** [reply]
I couldn't find an example, but here's how it should work:

```csharp
public class SampleWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .PublishMassTransitMessage(activity =>
            {
                activity.Set(a => a.Message, new MyMessage());
            });
    }
}
```

**Reply by pvanroos** [reply]
Thanks Sipke.  I figured it out.  I'm trying to coordinate all messaging in our system through these workflows built on top of Elsa.

**Reply by sfmskywalker** [reply]
Great! Please feel free to report any issues and or limitations you may encounter. Orchestrating message flow is an important aspect of Elsa, and something that we want to improve upon in V3, so any & all feedback is appreciated.
